### PR TITLE
Use workspace modules with `BP_GO_WORK_USE`

### PR DESCRIPTION
**Pull request by mamachanko**
## Summary

When a submodule uses `replace` directives referring to modules within the same repository using a relative path, e.g. `../../`, then the submodule cannot be built with paketo-buildpacks/go-build.

With Go's workspaces it's possible to include a submodule in the set of main modules so such a submodule can be built.

By allowing users to set `BP_GO_WORK_USE` the buildpack will first run `go work init` followed by `go work use <modules...>`, before building with `go build <...>`.

If this PR would be considered and ever get merged, I'd be happy to follow-up with a companion docs PR for https://paketo.io/docs/howto/go/.

## Use Cases

I want to build [cert-manager](https://github.com/cert-manager/cert-manager)’s images w/ buildpacks. Unfortunately, but to the best of my knowledge, as of `>=1.12` vanilla checkouts (e.g. [cert-manager@v.12.7](https://github.com/cert-manager/cert-manager/tree/v1.12.7)) cannot be built with [paketo-buildpacks/go@v4.6.1](https://github.com/paketo-buildpacks/go/tree/v4.6.1) for their use of relative `replace` directives. What I've tried:

- `path=./cmd/controller` doesn’t work b/c then it can’t see `../../`
- `BP_GO_TARGET=./cmd/controller` doesn’t work because it’s not contained by the main module
- `BP_GO_BUILD_FLAGS="-C ./cmd/controller"` doesn’t work b/c of a limitation (or by design) in paketo-buildpacks/go-build. _([it doesn’t put `-C` first](https://github.com/paketo-buildpacks/go-build/blob/main/go_build_process.go#L65), which `go build` expects. fixing it probably doesn’t help, b/c the buildpack makes assumptions about the working dir. it’s complicated)_

*(There’s a chance my research wasn’t exhaustive. Can you think of a way to build images off of vanilla cert-manager?)*

However, with Go workspaces a vanilla check-out of cert-manager _can_ be built. This is even documented by [cert-manager.io/building/#go-workspaces](http://cert-manager.io/building/#go-workspaces). When the workspace `use`s `./cmd/controller` it’s part of the main module and the buildpack succeeds. However, `go.work` is not being checked in by upstream. Afaik it's not common to check in `go.work*` either. As a workaround I _could_ augment every version of upstream I want to build with `go.work`, but that’s not ideal. I'd rather use build a vanilla checkout with buildpacks.

After building this branch and building paketo-buildpacks/go with it, cert-manager's images built.

```shell
pack build \
  cert-manager_controller:0.0.0-mamachanko \
  --builder paketobuildpacks/builder-jammy-tiny \
  --buildpack /Users/*****/workspace/go/build/buildpackage.cnb \
  --env BP_GO_WORK_USE="./cmd/controller" \
  --env BP_GO_TARGETS="./cmd/controller"
```

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by mamachanko** [comment]
@robdimsdale @ForestEckhardt Unfortunately, I couldn't figure out how to run integration tests locally. Reverse engineering the `.github/workflows/test-pull-request.yml` wasn't successful either since I couldn't figure out which `--builder` to pass to `./scripts/integration.sh`. I was hoping this PR would do that for me, but a semver label is missing.

**Review by ForestEckhardt — CHANGES_REQUESTED** [review]
Hey there! This PR look great. My only call out is that I would like to see some unit test for the new functionality that you added. I would be more than happy to consult with you on adding those or back filling them and letting you take a look. Just let me know!

**Comment by mamachanko** [comment]
> My only call out is that I would like to see some unit test for the new functionality that you added.

I shall happily add unit tests! 

Until then, would you be able to instruct me on how to run the integration locally?

File: go_build_process_test.go [hunk 1]
```diff
@@ -199,6 +199,66 @@ func testGoBuildProcess(t *testing.T, context spec.G, it spec.S) {
...
+
+			Expect(executable.ExecuteCall.Receives.Execution.Dir).To(Equal(workspacePath))
+			Expect(executable.ExecuteCall.Receives.Execution.Env).To(ContainElement(fmt.Sprintf("GOCACHE=%s", goCache)))
+
+			Expect(logs).To(ContainLines(
+				"  Executing build process",
+				"    Running 'go work init'",
+				"    Running 'go work use ./some/module1 ./some/module2'",
+				fmt.Sprintf(`    Running 'go build -o %s -buildmode pie -trimpath .'`, filepath.Join(layerPath, "bin")),
+				"      Completed in 0s",
```
Comment (mamachanko) [inline go_build_process_test.go:257]: 🤔 It's consistently expecting `0s` while other tests want `1s`. I can't quite tell why.

**Comment by mamachanko** [comment]
> It looks like all of the integration tests are failing at the moment. Could you take a look in that?

Done. Finally I figured out how to run the integration tests for a builder locally with:

```sh
./scripts/integration.sh --builder paketobuildpacks/builder-jammy-buildpackless-tiny --token "$GIT_TOKEN"
```

I understand that it runs across a matrix of `--builder` on CI, but I hope that tiny was a good-enough approximation locally.

Also, thanks to https://github.com/ForestEckhardt/freezer/issues/10 I was able to figure out how to rescue myself from `"failed to open cacheManager: EOF"` with `rm -rf $HOME/.freezer-cache`.

All that being said,

```sh
./scripts/unit.sh
./scripts/integration.sh --builder paketobuildpacks/builder-jammy-buildpackless-tiny --token "$GIT_TOKEN"
```

passes for me.

**Comment by ForestEckhardt** [comment]
I think that this look great!

**Review by ForestEckhardt — APPROVED** [review]

**Pull request merged**